### DISCUSSION TITLE
Example won't compile -  error: use of deleted function

### DISCUSSION THREAD

**Discussion by fscii** [pschatzmann/arduino-audio-tools#1579] [Q&A] [answered]
I did a git pull today, also installed the sd fat library today via git clone.

Using example: 
  https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-communication/a2dp/player-sdfat-a2dp

My ONLY change to code is cfg.name = "MyMusic";

Error Messages:

`In file included from C:\Users\fscii\AppData\Local\Temp\.arduinoIDE-unsaved2024428-17304-15na9t9.f6crf\player-sdfat-a2dp\player-sdfat-a2dp.ino:14:
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h: In member function 'virtual Stream* audio_tools::AudioSourceSDFAT::selectStream(const char*)':
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h:122:12: error: use of deleted function 'File32& File32::operator=(const File32&)'
     file = new_file;
            ^~~~~~~~
In file included from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatVolume.h:27,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatLib.h:28,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/SdFat.h:32,
                 from c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h:6,
                 from C:\Users\fscii\AppData\Local\Temp\.arduinoIDE-unsaved2024428-17304-15na9t9.f6crf\player-sdfat-a2dp\player-sdfat-a2dp.ino:14:
c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatFile.h:1088:7: note: 'File32& File32::operator=(const File32&)' is implicitly deleted because the default definition would be ill-formed:
 class File32 : public StreamFile<FatFile, uint32_t> {
       ^~~~~~
c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatFile.h:1088:7: error: use of deleted function 'StreamFile<FatFile, unsigned int>& StreamFile<FatFile, unsigned int>::operator=(const StreamFile<FatFile, unsigned int>&)'
In file included from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/ExFatLib/ExFatFile.h:883,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/ExFatLib/ExFatVolume.h:27,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/ExFatLib/ExFatLib.h:28,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/SdFat.h:31,
                 from c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h:6,
                 from C:\Users\fscii\AppData\Local\Temp\.arduinoIDE-unsaved2024428-17304-15na9t9.f6crf\player-sdfat-a2dp\player-sdfat-a2dp.ino:14:
c:\users\fscii\documents\arduino\libraries\sdfat\src\common\arduinofiles.h:61:7: note: 'StreamFile<FatFile, unsigned int>& StreamFile<FatFile, unsigned int>::operator=(const StreamFile<FatFile, unsigned int>&)' is implicitly deleted because the default definition would be ill-formed:
 class StreamFile : public stream_t, public BaseFile {
       ^~~~~~~~~~
c:\users\fscii\documents\arduino\libraries\sdfat\src\common\arduinofiles.h:61:7: error: 'FatFile& FatFile::operator=(const FatFile&)' is private within this context
In file included from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatVolume.h:27,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatLib.h:28,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/SdFat.h:32,
                 from c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h:6,
                 from C:\Users\fscii\AppData\Local\Temp\.arduinoIDE-unsaved2024428-17304-15na9t9.f6crf\player-sdfat-a2dp\player-sdfat-a2dp.ino:14:
c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatFile.h:158:12: note: declared private here
   FatFile& operator=(const FatFile& from) = default;
            ^~~~~~~~
In file included from c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h:12,
                 from C:\Users\fscii\AppData\Local\Temp\.arduinoIDE-unsaved2024428-17304-15na9t9.f6crf\player-sdfat-a2dp\player-sdfat-a2dp.ino:14:
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/SDDirect.h: In instantiation of 'void audio_tools::SDDirect<SDT, FileT>::listDir(const char*) [with SDT = SdFat32; FileT = File32]':
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/SDDirect.h:43:7:   required from 'const char* audio_tools::SDDirect<SDT, FileT>::operator[](int) [with SDT = SdFat32; FileT = File32]'
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h:105:36:   required from here
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/SDDirect.h:84:11: error: use of deleted function 'File32::File32(const File32&)'
       if (!isDirectory(root)) {
In file included from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatVolume.h:27,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatLib.h:28,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/SdFat.h:32,
                 from c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h:6,
                 from C:\Users\fscii\AppData\Local\Temp\.arduinoIDE-unsaved2024428-17304-15na9t9.f6crf\player-sdfat-a2dp\player-sdfat-a2dp.ino:14:
c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatFile.h:1088:7: note: 'File32::File32(const File32&)' is implicitly deleted because the default definition would be ill-formed:
 class File32 : public StreamFile<FatFile, uint32_t> {
       ^~~~~~
c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatFile.h:1088:7: error: use of deleted function 'StreamFile<FatFile, unsigned int>::StreamFile(const StreamFile<FatFile, unsigned int>&)'
In file included from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/ExFatLib/ExFatFile.h:883,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/ExFatLib/ExFatVolume.h:27,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/ExFatLib/ExFatLib.h:28,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/SdFat.h:31,
                 from c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h:6,
                 from C:\Users\fscii\AppData\Local\Temp\.arduinoIDE-unsaved2024428-17304-15na9t9.f6crf\player-sdfat-a2dp\player-sdfat-a2dp.ino:14:
c:\users\fscii\documents\arduino\libraries\sdfat\src\common\arduinofiles.h:61:7: note: 'StreamFile<FatFile, unsigned int>::StreamFile(const StreamFile<FatFile, unsigned int>&)' is implicitly deleted because the default definition would be ill-formed:
 class StreamFile : public stream_t, public BaseFile {
       ^~~~~~~~~~
c:\users\fscii\documents\arduino\libraries\sdfat\src\common\arduinofiles.h:61:7: error: 'constexpr FatFile::FatFile(const FatFile&)' is private within this context
In file included from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatVolume.h:27,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatLib.h:28,
                 from c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/SdFat.h:32,
                 from c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h:6,
                 from C:\Users\fscii\AppData\Local\Temp\.arduinoIDE-unsaved2024428-17304-15na9t9.f6crf\player-sdfat-a2dp\player-sdfat-a2dp.ino:14:
c:\Users\fscii\Documents\Arduino\libraries\SdFat\src/FatLib/FatFile.h:157:3: note: declared private here
   FatFile(const FatFile& from) = default;
   ^~~~~~~
In file included from c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioSourceSDFAT.h:12,
                 from C:\Users\fscii\AppData\Local\Temp\.arduinoIDE-unsaved2024428-17304-15na9t9.f6crf\player-sdfat-a2dp\player-sdfat-a2dp.ino:14:
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/SDDirect.h:137:10: note:   initializing argument 1 of 'bool audio_tools::SDDirect<SDT, FileT>::isDirectory(FileT) [with SDT = SdFat32; FileT = File32]'
     bool isDirectory(FileT f) {
          ^~~~~~~~~~~
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/SDDirect.h:94:7: error: use of deleted function 'File32::File32(const File32&)'
       if (isDirectory(root)){
       ^~
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/SDDirect.h:137:10: note:   initializing argument 1 of 'bool audio_tools::SDDirect<SDT, FileT>::isDirectory(FileT) [with SDT = SdFat32; FileT = File32]'
     bool isDirectory(FileT f) {
          ^~~~~~~~~~~
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/SDDirect.h:95:9: error: use of deleted function 'File32::File32(const File32&)'
         rewind(root);
         ^~~~~~
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/SDDirect.h:128:10: note:   initializing argument 1 of 'void audio_tools::SDDirect<SDT, FileT>::rewind(FileT) [with SDT = SdFat32; FileT = File32]'
     void rewind(FileT f){
          ^~~~~~
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/SDDirect.h:100:9: error: use of deleted function 'File32::File32(const File32&)'
         if (isDirectory(file)) {
         ^~
c:\Users\fscii\Documents\Arduino\libraries\arduino-audio-tools\src/AudioLibs/SDDirect.h:137:10: note:   initializing argument 1 of 'bool audio_tools::SDDirect<SDT, FileT>::isDirectory(FileT) [with SDT = SdFat32; FileT = File32]'
     bool isDirectory(FileT f) {
          ^~~~~~~~~~~

exit status 1

Compilation error: exit status 1`

**Comment by pschatzmann** [comment, marked as answer]
There seem to be some changes to the SDFAT library: If you read the Readme of that library you can find in the first sentence how to resolve this issue!

- set the FILE_COPY_CONSTRUCTOR_SELECT in the SdFatConfig.h to FILE_COPY_CONSTRUCTOR_PUBLIC or
- use the latest commit of the AudioTools library

**Comment by fscii** [comment]
I am using latest commit - it seems its been resolved.  Thank you!

There is one other issue.  If Audiokit is using SD card (several examples I've tried), I cannot have UART connected or I get errors like:

Example 1: Working SD
https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit/player-sd-audiokit
Warning: Pin '13' not set up because of conflict
However, it works via line out or speakers - even with warning.

Example 2: Not working SD
https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-communication/a2dp/player-sdfat-a2dp
Warning: Pin '13' not set up because of conflict

Except now SD card (same sd card, same path as example 1) isn't working?

[W] A2DPStream.h : 296 - ==> state: Connecting
[W] A2DPStream.h : 296 - ==> state: Connected
[E] AudioSourceSDFAT.h : 78 - sd.begin failed
[E] AudioSourceSDFAT.h : 110 - Filename is null
[W] AudioPlayer.h : 184 - -> begin: no data found

If I change headers and code to not use AudioSourceSDFAT.h anymore but make these changes:
#include "AudioLibs/AudioBoardStream.h"
#include "AudioLibs/AudioSourceSD.h"

AudioSourceSD source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);

I get similar results:
[W] A2DPStream.h : 296 - ==> state: Connecting
[W] A2DPStream.h : 296 - ==> state: Connected
[E] AudioSourceSD.h : 48 - SD.begin cs=13 failed

It seems both libraries are having an issue with the SD card in 2nd example but not in the first example.

What am I missing?

**Comment by pschatzmann** [comment]
The AudioKits have different pin conflicts. If you activate the SD some buttons will not work and it depends on the version which one it is. I suggest that you have a closer look at [pin assignments](https://github.com/pschatzmann/arduino-audio-driver/blob/main/src/DriverPins.h) of your device to get a better understanding.

The SD does not work in your second example because you did not activate it in your code! Have a look at the commented out code which shows how this is done for the AudioKit! The sketch uses the default (SPI and CS) pins of the ESP32 which are different then the ones needed by the AudioKit!

**Reply by Tanmay-J10** [reply]
Hi, big fan of yours, thanks for puuting in the efforrt  ;)

i have been having the same issue as above since a long time , but i am unable to solve it , could you help me pls
here is the log, could you help me pls


[I] AudioBoardStream.h : 60 - void audio_tools::AudioBoardStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 131 - ActionLogic::add pin: 36 / logic: 0
[I] AudioBoardStream.h : 60 - void audio_tools::AudioBoardStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 131 - ActionLogic::add pin: 23 / logic: 0
[I] AudioBoardStream.h : 60 - void audio_tools::AudioBoardStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 131 - ActionLogic::add pin: 19 / logic: 0
[I] AudioPlayer.h : 338 - setVolume(0.700000)
[I] VolumeStream.h : 199 - setVolume: 0.700000 at 0
[I] VolumeStream.h : 199 - setVolume: 0.700000 at 1
[I] AudioTypes.h : 128 - out: sample_rate: 44100 / channels: 2 / bits_per_sample: 16
[E] AudioSourceSD.h : 48 - SD.begin cs=13 failed
[E] AudioSourceSD.h : 48 - SD.begin cs=13 failed
[E] AudioSourceSD.h : 48 - SD.begin cs=13 failed



**_and here is the part of the code i updated_**


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  // sd_active is setting up SPI with the right SD pins by calling 
  SPI.begin(23, 19, 18, 5);
  cfg.sd_active = true;
  kit.begin(cfg);

  // setup additional buttons 
  kit.addDefaultActions();
  kit.addAction(kit.getKey(1), startStop);
  kit.addAction(kit.getKey(4), next);
  kit.addAction(kit.getKey(3), previous);


  // setup player
  player.setVolume(0.7);
  player.begin();

  // select file with setPath() or setIndex()
  player.setPath("/music/1.mp3");
  //player.setIndex(1); // 2nd file
  `
}

void loop() {
  player.copy();
  kit.processActions();

}

**Reply by pschatzmann** [reply]
Open a new discussion if something is not working: this is getting too confusing otherwise!

**Comment by fscii** [comment]
Surprise surprise, works fine lol.  I went through so many AudioKit examples, when I didn't find SD+BT Transmit, I found it in another folder and missed that I had to set all of the SPI active not just CS because I forgot it wasn't an AudioKit example.  Sorry about that and thanks!

**Comment by fscii** [comment]
There still seems to be one remaining issue.  2nd example, sd card only plays first file?  In 1st example, sd plays through all mp3 files.  ??

First song ending... then music stops and 2nd song not reached. -- AudioLogger set to Info

[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  680 -> 680 -> 680 bytes - in 1 hops
[I] AudioPlayer.h : 491 - -> timeout - moving by 1
[I] AudioPlayer.h : 505 - void audio_tools::AudioPlayer::writeEnd()

For reference, the sink/target is another ESP32 A2DP from ESP32_A2DP.

**Comment by pschatzmann** [comment]
Try to activate the A2DP logging. Maybe the break between the files is too long and A2DP hangs up because it could not read any data. If this is the case you can try to set cfg.silence_on_nodata = true;

**Comment by fscii** [comment]
On the source I have:
 AudioLogger::instance().begin(Serial, AudioLogger::Info);

Is there code you could share how to do A2DP logging? I haven't seen it yet.

I did anyway try cfg.silence_on_nodata = true; and I got the exact same error.
It has to be the 2nd example because the first example (BT sink) never disconnects from anything (pc, phone, etc)

**Comment by pschatzmann** [comment]
https://github.com/pschatzmann/ESP32-A2DP?tab=readme-ov-file#logging

If the SD library is working for you, why don't you just stick with it ?

**Comment by fscii** [comment]
I am trying to use BT as source from audiokit which has sd card.  This is why I'm using this example.  Has anyone else tried this example?

I added the additional logging set to Debug->Level: Info and here is the area of concern:  (using only a few, few second long mp3 sound files to speed up testing)  First one plays then it stops.  I still have  cfg.silence_on_nodata = true;

[ 10977][I][BluetoothA2DPSource.cpp:655] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state APP_AV_STATE_CONNECTED, evt 0x1
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  1024 -> 1024 -> 1024 bytes - in 1 hops
[I] StreamCopy.h : 147 - StreamCopy::copy  362 -> 362 -> 362 bytes - in 1 hops
[I] AudioPlayer.h : 491 - -> timeout - moving by 1
[I] AudioPlayer.h : 505 - void audio_tools::AudioPlayer::writeEnd()
[ 20554][I][BluetoothA2DPSource.cpp:655] bt_app_av_sm_hdlr(): [BT_AV] bt_app_av_sm_hdlr state APP_AV_STATE_CONNECTED, evt 0xff00

(this last line repeats forever)

Do you get the same result?

**Comment by pschatzmann** [comment]
Did you try to change your sketch to use the SD library ?

Maybe you need to be a little [bit more systematic](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working) to identfy your issue. This sketch is way too complicated to test the SDFAT library.
Did you alredy double check that the [I2S SDFAT version](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit/player-sdfat-audiokit) is working properly for you ?

**Comment by fscii** [comment]
I did try SD Library after SDFat gave some trouble.  Both work but both give same result.  I'm currently using the SD libary.
Here's the minor changes to library example I'm running.

Note: Running on AudioKit

`/**
 * @file player-sdfat-a2dp.ino
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-sdfat-a2dp/README.md
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */

//remove this, to find issues regarding mp3 decoding
#define HELIX_LOGGING_ACTIVE false

#include "AudioTools.h"
#include "AudioLibs/AudioBoardStream.h"  // not needed for AudioSourceSDFAT.h
#include "AudioLibs/A2DPStream.h"
#include "AudioLibs/AudioSourceSD.h"  // AudioSourceSDFAT.h
#include "AudioCodecs/CodecMP3Helix.h"

const char *startFilePath="/";
const char* ext="mp3";
//AudioSourceSDFAT source(startFilePath, ext); // , PIN_AUDIO_KIT_SD_CARD_CS);
AudioSourceSD source(startFilePath, ext, PIN_AUDIO_KIT_SD_CARD_CS);
A2DPStream out;
MP3DecoderHelix decoder;
AudioPlayer player(source, out, decoder);


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  // Just changed back to Info, had Warning prior to see more msgs

  // Setting up SPI if necessary with the right SD pins by calling 
  SPI.begin(PIN_AUDIO_KIT_SD_CARD_CLK, PIN_AUDIO_KIT_SD_CARD_MISO, PIN_AUDIO_KIT_SD_CARD_MOSI, PIN_AUDIO_KIT_SD_CARD_CS);

  // setup output - We send the test signal via A2DP - so we conect to a Bluetooth Speaker
  auto cfg = out.defaultConfig(TX_MODE);
  cfg.name = "MyMusic";  // set the device here. Otherwise the next available device is used for output
  // cfg.auto_reconnect = true;  // if this is use we just quickly connect to the last device ignoring cfg.name
  // cfg.sd_active = true;  // I added this but only works if Audiokit is the type?!

  cfg.silence_on_nodata = true;  //  suggested by author when a2dp disconnects after first file played

  out.begin(cfg);

  // setup player
  player.setVolume(0.5);
  player.begin();

}

void loop() {
  player.copy();
}
`
Both SD Examples work fine (when sink is speakers or line out)

https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit/player-sd-audiokit

https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit/player-sdfat-audiokit

**Comment by fscii** [comment]
For the record, I took your advice and looked at: 
https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working

I tried I2S microphone into A2DP --> out to another BT speaker (same one) and it works fine.
https://github.com/pschatzmann/arduino-audio-tools/blob/9045503daae3b21300ee7bb76c4ad95efe9e1e6c/examples/examples-communication/a2dp/basic-audiokit-a2dp/basic-audiokit-a2dp.ino

I just had to add:   a2dp_source.set_volume(64); because without volume control, horrific feedback probably due to 2 microphones on audiokit (?)

(New question.  why is BluetoothA2DPSource::set_volume(uint_8) but AudioPlayer::set_volume(float)?
Can all volume be made uint_8 and no float?

Still no luck on SD->A2DP tho.  Seems like something in SD breaks A2DP sink.

**Comment by pschatzmann** [comment]
BluetoothA2DPSource is part of the A2DP project and uses the original (unconverted) A2DP volume values.

The AudioTools project (classes) use volume values between 0 and 1.0 consistently.
